ethash/src/cache.rs

```diff
@@ -0,0 +1,310 @@
+// Copyright 2015-2017 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+// TODO: Mmap failing to build is transparent, we want to add a possibility to use RAM over a mmap
+//       for maximum speed (since the OS needs to zero the file when we do `set_len` - although I
+//       think it can occasionally prevent that from happening - and it has to page the data into
+//       the page cache from disk). I suggest an `OptimizeFor` struct that gets passed into an
```

Comment:
It would delay the PR more than it already has been, but I would be OK with doing that.